prove_test.go

```diff
@@ -938,3 +938,375 @@ func FuzzGetProofSubset(f *testing.F) {
 		}
 	})
 }
+
+func FuzzUndoProofChain(f *testing.F) {
+	var tests = []struct {
+		numAdds  uint32
+		duration uint32
+		seed     int64
+	}{
+		{3, 0x07, 0x07},
+	}
+	for _, test := range tests {
+		f.Add(test.numAdds, test.duration, test.seed)
+	}
+
+	f.Fuzz(func(t *testing.T, numAdds, duration uint32, seed int64) {
+		// simulate blocks with simchain
+		sc := newSimChainWithSeed(duration, seed)
+
+		p := NewAccumulator(true)
+		stump := Stump{}
+
+		undoData := []struct {
+			addHashes  []Hash
+			delHashes  []Hash
+			targets    []uint64
+			remembers  []uint32
+			updateData UpdateData
+			proof      Proof
+			roots      []Hash
+
+			blockHeight int
+		}{}
+		// We'll store the cached utxos here. This would be the equivalent
+		// of a wallet storing its own utxos.
+		var cachedProof Proof
+		var cachedHashes []Hash
+		for b := 0; b <= 25; b++ {
+			adds, duration, delHashes := sc.NextBlock(numAdds)
+
+			// The blockProof is the proof for the utxos being
+			// spent/deleted.
+			blockProof, err := p.Prove(delHashes)
+			if err != nil {
+				t.Fatalf("FuzzUndoProofChain fail at block %d. Error: %v", b, err)
+			}
+
+			// Sanity checking.
+			err = p.Verify(delHashes, blockProof)
+			if err != nil {
+				t.Fatal(err)
+			}
+
+			// Sanity checking.
+			for _, target := range blockProof.Targets {
+				n, _, _, err := p.getNode(target)
+				if err != nil {
+					t.Fatalf("FuzzUndoProofChain fail at block %d. Error: %v", b, err)
+				}
+				if n == nil {
+					t.Fatalf("FuzzUndoProofChain fail to read %d at block %d.", target, b)
+				}
+			}
+
+			// For logging.
+			origCachedHashes := make([]Hash, len(cachedHashes))
+			copy(origCachedHashes, cachedHashes)
+
+			// Cache new utxos that have a duration longer than 3.
+			var remembers []uint32
+			var newHashesToCache []Hash
+			for i, add := range adds {
+				if duration[i] > 3 {
+					remembers = append(remembers, uint32(i))
+					newHashesToCache = append(newHashesToCache, add.Hash)
+				}
+			}
+
+			// expectedCachedHashes is the expected hashes after the
+			// modify has happened.
+			expectedCachedHashes := make([]Hash, len(cachedHashes))
+			copy(expectedCachedHashes, cachedHashes)
+			expectedCachedHashes = removeHashes(expectedCachedHashes, delHashes, func(elem Hash) Hash { return elem })
+			expectedCachedHashes = append(expectedCachedHashes, newHashesToCache...)
+
+			// Update the proof with the deletions and additions that
+			// happen in this block.
+			rootHashes := make([]Hash, len(p.Roots))
+			for i := range rootHashes {
+				rootHashes[i] = p.Roots[i].data
+			}
+			addHashes := make([]Hash, len(adds))
+			for i := range addHashes {
+				addHashes[i] = adds[i].Hash
+			}
+
+			err = p.Modify(adds, delHashes, blockProof.Targets)
```

Comment:
Don't need this Modify & Undo here. This was added in the other test because I was trying to trigger a bug. No reason for this to be here.